### PR TITLE
Spark 3.4: Clarify concurrent table modification error in copy-on-write operations

### PR DESCRIPTION
**Pull request by aokolnychyi**
Tests that cover concurrent modifications and row-level operations sporadically fail for me locally. The primary reason is that if table caching is disabled, the table would be eagerly refreshed in each scan. That means we can pick up concurrent changes even though they were done via another `Table` instance.

### CODE REVIEW

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkCopyOnWriteScan.java [hunk 1]
```diff
@@ -107,6 +107,7 @@ public void filter(Filter[] filters) {
         Objects.equals(snapshotId(), currentSnapshotId()),
         "Runtime file filtering is not possible: the table has been concurrently modified. "
             + "Row-level operation scan snapshot ID: %s, current table snapshot ID: %s. "
+            + "If an external process modifies the table, enable table caching in the catalog. "
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkCopyOnWriteScan.java:110]: @rdblue, could you review this one? It is related to #6077 you took a look at earlier.

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java [hunk 2]
```diff
@@ -687,6 +696,13 @@ public synchronized void testUpdateWithSnapshotIsolation()
       throws InterruptedException, ExecutionException {
     // cannot run tests with concurrency for Hadoop tables without atomic renames
     Assume.assumeFalse(catalogName.equalsIgnoreCase("testhadoop"));
+    // if caching is off, the table is eagerly refreshed during runtime filtering
```
Comment (jerqi) [inline spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java:699]: Should we extract these code to a method in the `SparkRowLevelOperationsTestBase`?  If so, we can reduce the duplicated code and comment.

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java [hunk 3]
```diff
@@ -687,6 +696,13 @@ public synchronized void testUpdateWithSnapshotIsolation()
...
+    // this can cause a validation exception as concurrent changes would be visible
```
Comment (jerqi) [inline spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java:700]: Should we extract this code to a method in the `SparkRowLevelOperationsTestBase`? If so, we can reduce the duplicated code and comment.

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java [hunk 4]
```diff
@@ -948,6 +950,13 @@ public void testDeleteThatRequiresGroupingBeforeWrite() throws NoSuchTableExcept
   public synchronized void testDeleteWithSerializableIsolation() throws InterruptedException {
     // cannot run tests with concurrency for Hadoop tables without atomic renames
     Assume.assumeFalse(catalogName.equalsIgnoreCase("testhadoop"));
+    // if caching is off, the table is eagerly refreshed during runtime filtering
+    // this can cause a validation exception as concurrent changes would be visible
+    Assume.assumeTrue(
+        PropertyUtil.propertyAsBoolean(
+            catalogConfig,
+            CatalogProperties.CACHE_ENABLED,
+            CatalogProperties.CACHE_ENABLED_DEFAULT));
```
Comment (szehon-ho) [inline spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java:959]: Nit: how about making this explicitly true?

**Review by szehon-ho — APPROVED** [review]

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java (hunk 3 above)
Comment (aokolnychyi) [reply spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java:694]: I added a common method to check whether caching is enabled, but I feel the comment must be in the test itself for better visibility.

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java (hunk 4 above)
Comment (aokolnychyi) [reply spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java]: We have to use the actual default value cause we are simply checking if it is enabled. If we later change the default value, this test should not be affected.

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @szehon-ho @jerqi!

**Pull request merged**